xoinvader/animation.py

```diff
@@ -0,0 +1,101 @@
+"""Animation.
+
+Animation is set of keyframes.
+Value of selected attribute changes in time.
+
+Keyframe:
+  (time, value)
+
+Objects have animation manager which manages animation graph and switching."""
+
+
+from operator import itemgetter
+
+from xoinvader.utils import Timer
+
+
+# TODO: Implement animation graph and etc
+class AnimationManager(object):
+    """Manage list of object animation."""
+
+    def __init__(self):
+        self._animations = {}
+        self._current = None
+
+    def bind(self, name, obj):
+        """Transfer animation to other object."""
+        self._animations[name].bind(obj)
+
+    def add(self, name, *args, **kwargs):
+        """Add new animation, pass args to Animation class.
+
+        :param str name: animation name
+        """
+
+        self._animations[name] = Animation(name, *args, **kwargs)
+        if len(self._animations) == 1:
+            self._current = name
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        try:
+            next(self._animations[self._current])
+        except StopIteration:
+            raise  # We need to change animation
+
+
+class Animation(object):
+    """Animation unit.
+
+    :param str name: animation name
+    :param object bind: object to bind animation
+    :param str attr: attribute to change in frames
+    :param list keyframes: {float: object} dicts
+    :param bool loop: loop animation or not
+    """
+
+    def __init__(self, name, bind, attr, keyframes, loop=False):
+        self._name = name
+        self._obj = bind
+        self._attr = attr
+        self._keyframes = sorted(keyframes, key=itemgetter(0))
+        self._loop = loop
+        if not keyframes:
+            raise ValueError("Animation keyframes must not be empty.")
+
+        self._timer = Timer(self._keyframes[-1][0], lambda: True)
+        self._timer.start()
+
+        self._current = 0
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+
+        if not self._timer.running:
+            self._timer.restart()
+
+        self._timer.update()
+
+        # TODO: feature-animation: approximation
+        if len(self._keyframes) == self._current:
+            if self._loop:
+                self._current = 0
+                self._timer.stop()
+            else:
+                self._timer.stop()
+                raise StopIteration
+
+        keyframe = self._keyframes[self._current]
+        if self._timer.get_elapsed() >= keyframe[0]:
```

Comment:
Will be, when all will be done. Now we just pick next keyframe when previous expired. No interpolation, no complicated logic.